### ISSUE TITLE
Test Database Migration and Rollback

### ISSUE THREAD

**Issue by Pcolar** [dpn-admin/dpn-server#110]
Based on the current production deployment, create a test instance that rolls a database instance forward from v1.x to v2.00. The test environment should also demonstrate rollback capability.

If rollback cannot be achieved for specific migration steps, document and test possible recovery scenarios.

**Comment by dazza-codes** [comment]
I have a production MySQL dump from TDR for testing the API v2 migrations.

**Comment by dazza-codes** [comment]
When the TDR prod dump is loaded and the migration status is checked, we get a clean status report that everything is migrated up to and including `20151215211317` migration, i.e.

```
   up     20151003181949  Remove restore statuses table
   up     20151019220704  Remove replication statuses table
   up     20151215211317  Reverse replication transfer bag man request association
  down    20160222203129  Refactor replication transfer
  down    20160226014529  Change bag man requests
  down    20160226193341  Add unpack location to bag man requests
  down    20160226213102  Add last error to bag man requests
  down    20160429211031  Change fixity checks to digests
  down    20160429211032  Add bags to digests
  down    20160429211033  Add nodes to message digests
  down    20160429211034  Add timestamps to message digests
  down    20160510205152  Create table fixity checks
  down    20160510205153  Add bags to fixity checks
  down    20160510205154  Add nodes to fixity checks
  down    20160616191514  Create ingests
  down    20160701185802  Create run times
  down    20160803214414  Delete table delayed jobs
  down    20160805220611  Change table restore transfers
  down    20160805221503  Disallow null on replication
  down    20160805224340  Allow null on restore transfer link
  down    20160811162500  Change member uuid to member
  down    20160921184937  Add cancel reason detail to replication transfer
  down    20160921185130  Add cancel reason detail to restore transfer
  down    20160927230134  Disable fk message digests nodes
  down    20160927230135  Node not null in message digest
  down    20160927230136  Enable fk message digests nodes
```

**Comment by dazza-codes** [comment]
There is at least one error in the migrations, e.g.

```
 RAILS_ENV=production bundle exec rake db:migrate
== 20160222203129 RefactorReplicationTransfer: migrating ======================
-- add_column(:replication_transfers, :store_requested, :boolean, {:null=>false, :default=>false})
   -> 0.6996s
-- add_column(:replication_transfers, :stored, :boolean, {:null=>false, :default=>false})
   -> 0.7167s
-- add_column(:replication_transfers, :cancelled, :boolean, {:null=>false, :default=>false})
   -> 0.6949s
-- add_column(:replication_transfers, :cancel_reason, :text, {:default=>nil})
   -> 0.7199s
-- remove_column(:replication_transfers, :fixity_accept)
   -> 0.6808s
-- remove_column(:replication_transfers, :bag_valid)
   -> 0.6928s
-- remove_column(:replication_transfers, :status)
   -> 0.7089s
== 20160222203129 RefactorReplicationTransfer: migrated (5.1285s) =============

== 20160226014529 ChangeBagManRequests: migrating =============================
-- add_column(:bag_man_requests, :cancel_reason, :text, {:default=>nil})
   -> 0.2608s
-- rename_column(:bag_man_requests, :status, :last_step_completed)
   -> 0.0047s
-- remove_column(:bag_man_requests, :validity)
   -> 0.2364s
== 20160226014529 ChangeBagManRequests: migrated (0.5405s) ====================

== 20160226193341 AddUnpackLocationToBagManRequests: migrating ================
-- add_column(:bag_man_requests, :unpacked_location, :string)
   -> 0.2667s
== 20160226193341 AddUnpackLocationToBagManRequests: migrated (0.2668s) =======

== 20160226213102 AddLastErrorToBagManRequests: migrating =====================
-- add_column(:bag_man_requests, :last_error, :text)
   -> 0.2324s
== 20160226213102 AddLastErrorToBagManRequests: migrated (0.2326s) ============

== 20160429211031 ChangeFixityChecksToDigests: migrating ======================
-- remove_foreign_key(:fixity_checks, :bags)
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Table 'fixity_checks' has no foreign key on column 'bag_id'
```

**Comment by dazza-codes** [comment]
It _might_ be feasible to modify an existing migration by adding a condition that tests for the existence of an FK before trying to remove it, something like
http://stackoverflow.com/questions/21795023/how-to-create-a-migration-to-remove-an-index-only-if-it-exists-rather-than-thro

(In general, it's not wise to modify an existing migration that might have already run.  But if it's just adding a condition on removing the FK, it might be OK.)

**Comment by dazza-codes** [comment]
Removing a foreign key might be wrapped in a check for the existence of the FK, something like:

``` ruby
fk_column = foreign_key_column_for(options_or_to_table)

if foreign_keys(from_table).include? fk_column
  #remove_foreign_key
end
```

Based on checking the source code details at
http://apidock.com/rails/v4.2.7/ActiveRecord/ConnectionAdapters/SchemaStatements/remove_foreign_key

**Comment by dazza-codes** [comment]
The check for an FK (noted above) fixed that migration.  Another error appears in a later migration, i.e.

```
== 20160429211034 AddTimestampsToMessageDigests: migrating ====================
-- add_column(:message_digests, :created_at, :datetime)
   -> 0.3097s
-- change_column(:message_digests, :created_at, :datetime, {:null=>false})
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: Invalid use of NULL value: ALTER TABLE `message_digests` CHANGE `created_at` `created_at` datetime NOT NULL
```

**Comment by dazza-codes** [comment]
Migration fixes in the `develop-mysql-migration-fixes` branch are working on the TDR and Chron prod-db dumps.  But there is a piece of code in the revised migrations that might not work on a Postgres system, so that needs to be checked.

**Comment by dazza-codes** [comment]
This issue should be resolved now by the work merged from #119 (and #120).
